### ISSUE TITLE
Orchestrator mariadb failover blocked

### ISSUE THREAD

**Issue by rlittle316** [openark/orchestrator#1296]
We have a 3 node mariadb cluster with orchestrator running on all 3 nodes in a raft configuration. Recently we had an issue where one node in the cluster was in an amazon availability zone that had an outage. In this case the node in that AZ was the mysql master and the orchestrator leader. We see that orchestrator failed over to a different node as expected, but it did not ever do a mysql failover. We see the following error message repeating a few times that seems to be the culprit but we dont see any evidence that a failover was attempted before this in a log or on the db. Can anyone provide any guidance on how we ended up in this state, and how to troubleshoot?

`ERROR AttemptRecoveryRegistration: cluster NODE1 has recently experienced a failover (of NODE1) and is in active period. It will not be failed over again. You may acknowledge the failure on this cluster (-c ack-cluster-recoveries) or on NODE1 (-c ack-instance-recoveries) to remove this blockage`

**Comment by rlittle316** [comment]
I was able to reproduce the issue by putting the mysql, orchestrator and raft leader on the same node then simulating the az outage by blocking the network from that node.

**Comment by shlomi-noach** [comment]
It seems like a failover actually started, and the leader orchestrator node was able to propagate that information to the other raft members nodes, before it lost contact with the raft group. This is why the other nodes hesitate to run a failover.

When I say "a failover actually started" it may be just the first immediate step, which is to say "I'm going to run a failover" -- this is why you may not find evidence on the MySQL servers themselves.

It is weird that a network failure will cause this situation. It takes a couple seconds to detect a failure. During those 2 seconds, surely the orchestrator leader would lose touch with the other raft members. So I don't understand this part.

**Comment by rlittle316** [comment]
The funny thing is, while it looks like raft moved to a different node, each node claims to be its own cluster after this event and there is no replica for any orchestrator nodes. It cannot failover again in this state and the db has no master. Do you have any suggestions for troubleshooting this behavior?

**Comment by shlomi-noach** [comment]
> each node claims to be its own cluster

The reuse of "node" is ambiguous. Did you mean MySQL cluster?

**Comment by rlittle316** [comment]
Let me rephrase to be more clear. Raft moved to a different server in the cluster. Orchestrator is reporting that each server is its own cluster at that point and that there are no orchestrator replicas. Of the 3 server cluster, 1 server is down entirely, 2 servers are running orchestrator and both seem to have no replicas. 1 of those 2 is the raft leader. The 2 remaining servers have mariadb still running but no master. It seems because the orchestrator failover was not graceful (and arguably not successful) the db did not failover. 

I was trying to get orchestrator on the server that was not reporting as the raft leader to become a follower of the other server but have yet to find a way to do that. My hunch being that because orchestrator is in this state, it is not handling the db failover properly. 

Expected result after fail over in 3 server cluster: 

- orchestrator Leader server
    - orchestrator Follower server 
DB is failed over to new master and we have 1 master and 1 follower

actual result in this case:

- Leader server
- Leader server
DB is not failed over and we have 2 nodes set to be followers without a master.

Please note we are using orchestrator-3.1.4-1.x86_64.rpm.

I hope this is a bit more clear. I know this is alot going on. Please let me know if you have more questions.

**Comment by shlomi-noach** [comment]
> Let me rephrase to be more clear. Raft moved to a different server in the cluster. Orchestrator is reporting that each server is its own cluster at that point and that there are no orchestrator replicas. Of the 3 server cluster, 1 server is down entirely, 2 servers are running orchestrator and both seem to have no replicas. 1 of those 2 is the raft leader. The 2 remaining servers have mariadb still running but no master. It seems because the orchestrator failover was not graceful (and arguably not successful) the db did not failover.

Right. That's in line with the analysis of the situation thus far. The two replicas are "orphaned" until a new primary is promoted.

> I was trying to get orchestrator on the server that was not reporting as the raft leader to become a follower of the other server but have yet to find a way to do that.

That should be strictly automatic. The node should rejoin the cluster. Question: did it perhaps change IP address while it was down?

> Leader server
> Leader server

Uhhh, that's the first I ever hear about this situation. Two different orchestrator/raft nodes both claim to be Leader at the same time? If so, I have no idea nor suggestion. The raft protocol should never allows such a situation.

**Comment by rlittle316** [comment]
No servers IP addresses were changed in the process. There was only 1 server that was taken out via AZ outage, or in the test case I am using a firewall to block the other nodes from that one to simulate an AZ outage ( node 1 let say). That server was the Mariadb, Orchestrator, and raft leader.  It looks like when the followers could not connect to the original leader they did the election to move the raft to a new server (node 2 lets say). That went through, but when I look at the orchestrator topology it shows each server is its own master, and none have a replica. This seems like it is not expected, but I have not been able to come up with why it was unable to gracefully move orchestrator leader to a new server and keep the replicas from the other server and fail over properly. Is there anything I can provide that might be useful?

**Comment by shlomi-noach** [comment]
> but I have not been able to come up with why it was unable to gracefully move orchestrator leader to a new server 

I'm confused. I completely understand the description about the MariaDB servers each becoming their own cluster. But I'm confused about "move orchestrator leader". This conflicts with "they did the election to move the raft to a new server (node 2 lets say)".

So I'd like to clarify please, did the orchestrator raft leader correcly change into "node 2", and did the original leader lose leadership, or did you end up with two orchestrator raft leaders?

**Comment by rlittle316** [comment]
I was under the impression that the Mariadb, Orchestrator, and raft leaders were all separate. I am running `orchestrator-client -c all-cluster-masters` to see which node is the orchestrator leader before and after I manually take out one server. I also run `orchestrator-client -c raft-leader-hostname` to see which node is the raft leader. I then check the db to see what node is master. Before taking 1 of the servers down I verify all 3 leaders are on the same server. Then on that server I am blocking the other 2 server IP addresses. Once this is done I am expecting orchestrator to failover raft, mariadb, and orchestrator leaders to one of the 2 servers that can still communicate with each other. What is actually happening is raft fails over to one of the remaining servers, but Mariadb does not. When I run the all-cluster-masters command at this point I get 2 masters, the original and a new one( which is one of the remaining servers). I then check the db and the db is reporting to have no master. When I run `topology-tabulated on the new orchestrator master, it only shows itself in the cluster and has no reference to the other server that should be in the cluster still. When I run the same command on the old master It shows the only server is itself as well, but it is marked downtimed, with unknown and invalid in the status. I then ran the command on the server that is not reporting in the other 2 and see that it is ok but not related to any other servers. At this point the original server that was the leader is still blocked so these remaining 2 servers cannot communicate with it, and the db is reporting having no master. When I connect to the individual nodes they seem to be all working alone and have no replication running.

**Comment by shlomi-noach** [comment]
> I was under the impression that the Mariadb, Orchestrator, and raft leaders were all separate.

MariaDB and `orchestrator` are separate, and it just so happens that you run both on same machine. That's fine.

`raft` is a property of `orchestrator`. You have configured `orchestrator` to run with `raft` consensus, which means the orchestrator leader is the raft leader.

Perhaps you could paste your `orchestrator.conf.json` file?

> I am running orchestrator-client -c all-cluster-masters to see which node is the orchestrator leader before and after

That doesn't tell you anything about the orchestrator leader. That tells you about the database replication servers.

Hmmm... it occurs to me that maybe your confusion is that you have a MySQL/MariaDB replication topology as the backend for your `orchestrator` setup? If that is the case, that's the wrong way of doing it. Please see https://github.com/openark/orchestrator/blob/master/docs/high-availability.md

**Comment by rlittle316** [comment]
Thank you for that explanation about the commands. I was clearly off about my understanding of the outputs. It sounds like since raft is failing over to another node and says its healthy, the issue is really around the mariadb cluster failing over properly.

Based on the config file it seems we are using sqllite backend. Does this look like we configured it the expected way? BTW i hid some ip addresses and pws.

```
cat orchestrator.conf.json
{
  "Debug": false,
  "ListenAddress": ":3308",
  "MySQLTopologyUser": "orchestrator",
  "MySQLTopologyPassword": "HIDDEN",
  "MySQLTopologySSLSkipVerify": true,
  "BackendDB": "sqlite",
  "SQLite3DataFile": "/usr/local/orchestrator/orchestrator.sqlite3",
  "MySQLConnectTimeoutSeconds": 1,
  "DefaultInstancePort": 3307,
  "DiscoverByShowSlaveHosts": true,
  "HostnameResolveMethod": "default",
  "MySQLHostnameResolveMethod": "@@report_host",
  "SlaveLagQuery": "",
  "DetectClusterAliasQuery": "SELECT 'eng-test-tester'",
  "DetectClusterDomainQuery": "",
  "DetectInstanceAliasQuery": "SELECT SUBSTRING_INDEX(@@hostname, '.', 1)",
  "DetectSemiSyncEnforcedQuery": "",
  "StatusEndpoint": "/api/status",
  "StatusSimpleHealth": true,
  "RecoveryPeriodBlockSeconds": 5,
  "RecoverMasterClusterFilters": [
    "*"
  ],
  "RecoverIntermediateMasterClusterFilters": [
    "*"
  ],
  "OnFailureDetectionProcesses": [
    "echo $(date -u \"+%F %T %Z\") - 'Detected {failureType} on {failureCluster}. Affected replicas: {countSlaves}' >> /usr/local/appian/orchestrator/logs/orchestrator-recovery.log"
  ],
  "PreFailoverProcesses": [
    "echo $(date -u \"+%F %T %Z\") - 'Will recover from {failureType} on {failureCluster}' >> /usr/local/appian/orchestrator/logs/orchestrator-recovery.log"
  ],
  "PostFailoverProcesses": [
    "echo $(date -u \"+%F %T %Z\") - '(for all types) Recovered from {failureType} on {failureCluster}. Failed: {failedHost}:{failedPort}; Successor: {successorHost}:{successorPort}' >> /usr/local/appian/orchestrator/logs/orchestrator-recovery.log",
    "/usr/local/appian/mysql-failover/after-failover.sh {failedHost} {failedPort} {successorHost} {successorPort}"
  ],
  "PostUnsuccessfulFailoverProcesses": [
    "echo $(date -u \"+%F %T %Z\") - 'Unsuccessful failover process for {failureType} on {failureCluster}' >> /usr/local/appian/orchestrator/logs/orchestrator-recovery.log"
  ],
  "PostMasterFailoverProcesses": [
    "echo $(date -u \"+%F %T %Z\") - 'Recovered from {failureType} on {failureCluster}. Failed: {failedHost}:{failedPort}; Promoted: {successorHost}:{successorPort}' >> /usr/local/appian/orchestrator/logs/orchestrator-recovery.log"
  ],
  "PostIntermediateMasterFailoverProcesses": [
    "echo $(date -u \"+%F %T %Z\") - 'Recovered from {failureType} on {failureCluster}. Failed: {failedHost}:{failedPort}; Successor: {successorHost}:{successorPort}' >> /usr/local/appian/orchestrator/logs/orchestrator-recovery.log"
  ],
  "RaftEnabled": true,
  "RaftDataDir": "/var/lib/orchestrator",
  "RaftBind": "NODE3",
  "DefaultRaftPort": 3309,
  "RaftNodes": [
    "NODE1", "NODE2", "NODE3"
  ],
  "GraphitePollSeconds": 60,
  "GraphiteAddr": "localhost:9109",
  "GraphitePath": "orchestrator"
}
```

**Comment by shlomi-noach** [comment]
Yep, config looks good. You're running `raft` setup using SQLite backend. So there's no MySQL/MariaDB backend for orchestrator.

So we have a terminology issue which I'm not sure how to resolve:

> Before taking 1 of the servers down I verify all 3 leaders are on the same server.

I'm not sure how three leaders would be on the same server? Do you mean like you have three different MariaDB topologies and it just so happens that all primaries are co-located on same server. I'm really confused.

>  I am expecting orchestrator to failover raft, mariadb, and orchestrator leaders to one of the 2 servers that can still communicate with each other.

orchestrator/raft will pick a leader out of the two remaining servers, and, assuming they're co-located on MariaDB servers (which is allowed, but confusing), one of the MariaDB servers should be promoted as primary (`master`). I understand that the latter does not happen, but the description is confusing.

> When I run `topology-tabulated on the new orchestrator master, it only shows itself in the cluster and has no reference to the other server that should be in the cluster still. 

Very confused. Why should `orchestrator` show itself in `topology-tabulated`? `orchestrator` isn't MySQL. Plus it uses SQLite as backend.

Current summary is that we're having a broken discussion...

**Comment by rlittle316** [comment]
We have 3 servers and each one has orchestrator and mariadb on them. 

When I was referencing the 3 leaders i meant mariadb, orchestrator and raft because I still thought these were separate at that time. You have cleared up that the raft leader is truthfully the orchestrator leader as well, so that was my confusion in the first place. Now that I also know that the command `orchestrator-client -c all-cluster-masters` is returning information about the mariadb cluster and not specifically orchestrator, it makes sense that each node is only returning itself in `topology-tabulated` since this is also about the mariadb cluster and not about orchestrator. 

With this in mind, it looks like what happens is orchestrator/raft failed over to one of the remaining servers without issue, but mariadb did not. Does this help clear up the discussion? The question still remains, why did orchestrator not recover the db after raft failed over?

**Comment by shlomi-noach** [comment]
> With this in mind, it looks like what happens is orchestrator/raft failed over to one of the remaining servers without issue, but mariadb did not. Does this help clear up the discussion? 

Yes, that makes sense and that was my original interpretation of the problem.

> The question still remains, why did orchestrator not recover the db after raft failed over?

Therefore my original comment: https://github.com/openark/orchestrator/issues/1296#issuecomment-814617402 is the best description I can think of. And I don't have a good understanding of why/how the situation would happen, but I think that's what's hapenning.

**Comment by rlittle316** [comment]
I did find evidence that one of the 2 nodes had the change master to command executed. It looks like this node was going to be the master as it was set up as if it would be the new master. The other available node looks like it never received this change of master command. I actually just tested running the change of master command myself and mariadb is showing that its online with a master and slave. Is there any way we can ensure that the failover process is run on both nodes or troubleshoot why they may not have been? Even more information about what recovery steps were taken on a given node would be helpful I Imagine, but I am not sure how to investigate that.
